receiver/loadgenreceiver/logs.go

```diff
@@ -0,0 +1,151 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package loadgenreceiver // import "github.com/elastic/opentelemetry-collector-components/receiver/loadgenreceiver"
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	_ "embed"
+	"os"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/plog"
+	"go.opentelemetry.io/collector/receiver"
+	"go.uber.org/zap"
+)
+
+//go:embed testdata/logs.jsonl
+var demoLogs []byte
+
+type receiverLogs struct {
+	logs     plog.Logs
+	jsonSize int
+}
+
+type logsGenerator struct {
+	cfg    *Config
+	logger *zap.Logger
+
+	sampleLogs      []receiverLogs
+	lastSampleIndex int
+	consumer        consumer.Logs
+
+	cancelFn context.CancelFunc
+}
+
+func createLogsReceiver(
+	ctx context.Context,
+	set receiver.Settings,
+	config component.Config,
+	consumer consumer.Logs,
+) (receiver.Logs, error) {
+	genConfig := config.(*Config)
+	recv := logsGenerator{
+		cfg:             genConfig,
+		logger:          set.Logger,
+		consumer:        consumer,
+		sampleLogs:      make([]receiverLogs, 0),
+		lastSampleIndex: 0,
+	}
+
+	parser := plog.JSONUnmarshaler{}
+	var err error
+	sampleLogs := demoLogs
+
+	if genConfig.Logs.JsonlFile != "" {
+		sampleLogs, err = os.ReadFile(string(genConfig.Logs.JsonlFile))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	scanner := bufio.NewScanner(bytes.NewReader(sampleLogs))
+	for scanner.Scan() {
+		logBytes := scanner.Bytes()
+		lineLogs, err := parser.UnmarshalLogs(logBytes)
+		if err != nil {
+			return nil, err
+		}
+		recv.sampleLogs = append(recv.sampleLogs, receiverLogs{
+			logs:     lineLogs,
+			jsonSize: len(logBytes),
+		})
+	}
+
+	return &recv, nil
+}
+
+func (ar *logsGenerator) Start(ctx context.Context, _ component.Host) error {
+	startCtx, cancelFn := context.WithCancel(ctx)
+	ar.cancelFn = cancelFn
+
+	go func() {
+		for {
+			select {
+			case <-startCtx.Done():
+				return
+			default:
+				nMetrics, _, err := ar.nextLogs()
+				if err != nil {
+					ar.logger.Error(err.Error())
+					continue
+				}
+				err = ar.consumer.ConsumeLogs(startCtx, nMetrics)
```

Comment:
qq Will this method block once the configured rate limit is reached?

Reply:
It will not. It keeps erroring out with too many requests (what apm-server otlp returns):
```
2025-01-09T15:32:17.362Z	error	loadgenreceiver/metrics.go:114	too many requests	{"kind": "receiver", "name": "loadgen", "data_type": "metrics"}
github.com/elastic/opentelemetry-collector-components/receiver/loadgenreceiver.(*metricsGenerator).Start.func1
	/home/carson/projects/opentelemetry-collector-components/receiver/loadgenreceiver/metrics.go:114
```

Reply:
We could add a little bit of backoff for "too many requests" specifically to avoid hogging the cpu.

Reply:
the too many requests is actually from ratelimiter. Might need to make ratelimiter configurable, whether to block or to return immediately.

Reply:
Disabling ratelimitprocessor for now. We can improve ratelimitprocessor later.

Reply:
> the too many requests is actually from ratelimiter. Might need to make ratelimiter configurable, whether to block or to return immediately. I haven't got any 429 from apm-server yet.

(for future) What about wrapping it under an exponential backoff strategy (e.g https://github.com/cenkalti/backoff?tab=readme-ov-file)? (to auto determine an acceptable ratelimit)

Reply:
The idea behind the exponential backoff would be to make the `loadgen` receiver yield the cpu once it reaches the limit, thus, making the `delay` sleep on the receiver. Currently, if a low rate limit is set, the receiver will overwhelm the logs with:
```
2025-01-13T19:23:36.418+0100    error   loadgenreceiver@v0.0.0/metrics.go:99    too many requests       {"kind": "receiver", "name": "loadgen", "data_type": "metrics"}
2025-01-13T19:23:36.419+0100    error   loadgenreceiver@v0.0.0/metrics.go:99    too many requests       {"kind": "receiver", "name": "loadgen", "data_type": "metrics"}
```

But my understanding is that we are testing on a known rate limit at the moment, so I agree on adding this feature later.

Reply:
> can you retry after pulling in the latest changes of this PR?

Just pulled the repository and updated the `loadgen/config.example.yaml` with this rate limit configuration:

```yaml
  ratelimit:
    strategy: records
    rate: 5
    burst: 5
    throttle_behavior: delay
```

`$ ELASTIC_APM_SERVER_URL=http://localhost:8200 ELASTIC_APM_API_KEY=some_api_key ./_build/elastic-collector-components --config loadgen/config.example.yaml`

And a big amount of logs with "too many requests" are still being generated. As mentioned, if we do not plan to work with this low rate limits, I am fine with postponing this feature

Reply:
sorry I spoke too soon. The "too many requests" are no longer there at 5000, but are still there at 5. Let me check why.

Reply:
https://pkg.go.dev/golang.org/x/time/rate#Limiter.ReserveN
> The returned Reservation’s OK() method returns false if n exceeds the Limiter's burst size.

I think we can skip this edge case first. This seems to be working as expected. If it only allows 5 item burst, you can never send 6 items in 1 go.